xonsh/environ.py

```diff
@@ -82,21 +97,17 @@ def format_prompt(template=default_prompt):
     + NO_COLOR -- Resets any previously used color codes
     """
     env = builtins.__xonsh_env__
-    cwd = env['PWD']
-    branch = current_branch(cwd=cwd)
-    branch = '' if branch is None else ' ' + branch
-    p = template.format(
-            user=env.get('USER', '<user>'),
-            hostname=socket.gethostname(),
-            cwd=cwd.replace(env['HOME'], '~'),
-            curr_branch=branch,
-            **TERM_COLORS
-            )
-    return p
+    template = template() if callable(template) else template
+    fmt = env.get('FORMAT_DICT', FORMAT_DICT)
+    included_names = set(i[1] for i in FORMATTER.parse(template))
+    fmt = {k: v for (k, v) in fmt.items() if k in included_names}
+    fmt = {k: (v() if callable(v) else v) for (k, v) in fmt.items()}
```

Comment:
This might be faster as a single loop over fmt, though this is certainly the most expressive. Probably doesn't matter too much.